#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
ä¸»çª—å£ç•Œé¢ - ä½¿ç”¨ Qt Designer UI æ–‡ä»¶
åŸºäº usb_manager.ui ç”Ÿæˆï¼Œå¯åœ¨ Designer ä¸­å¯è§†åŒ–ç¼–è¾‘
"""

import getpass
from pathlib import Path
from PyQt6.QtWidgets import QMainWindow, QTableWidgetItem, QFileDialog, QMessageBox, QPushButton, QHeaderView
from PyQt6.QtCore import Qt, QTimer
from PyQt6.QtGui import QFont

from .usb_manager_ui import Ui_MainWindow
from ..core.usb_scanner import USBScanner
from ..core.drive_manager import DriveManager
from ..core.file_transfer import FileTransferThread
from .styles import AppStyles


class USBManagerWindow(QMainWindow):
    """USB è®¾å¤‡ç®¡ç†å™¨ä¸»çª—å£ - ä½¿ç”¨ UI æ–‡ä»¶ç‰ˆæœ¬"""
    
    def __init__(self):
        super().__init__()
        
        # åŠ è½½ UI
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)
        
        # æ•°æ®
        self.selected_drive = None
        self.transfer_thread = None
        
        # åº”ç”¨æ ·å¼
        self.apply_styles()
        
        # è¿æ¥ä¿¡å·
        self.connect_signals()
        
        # æ›´æ–°ç”¨æˆ·ä¿¡æ¯
        self.ui.userLabel.setText(f"ğŸ‘¤ ç”¨æˆ·: {getpass.getuser()}")
        
        # å¯åŠ¨å®šæ—¶å™¨ - Windowsä¸Šæ”¹ä¸º10ç§’ï¼Œé¿å…å¡é¡¿
        self.timer = QTimer()
        self.timer.timeout.connect(self.auto_refresh)
        self.timer.start(10000)  # æ¯10ç§’åˆ·æ–°ï¼ˆé¿å…é¢‘ç¹æ‰«æå¯¼è‡´å¡é¡¿ï¼‰
        
        # åˆå§‹åŠ è½½
        self.refresh_all()
    
    def apply_styles(self):
        """åº”ç”¨æœ€å°æ ·å¼ - åªè®¾ç½®å¿…è¦çš„åŠŸèƒ½ï¼Œä¿æŒ UI æ–‡ä»¶åŸæ ·"""
        # æ›´æ–°ç”¨æˆ·æ ‡ç­¾æ–‡å­—é¢œè‰²ä½¿å…¶åœ¨æ ‡é¢˜æ ä¸­å¯è§
        self.ui.headerWidget.setStyleSheet(f"""
            QWidget {{
                background: qlineargradient(x1:0, y1:0, x2:1, y2:0,
                    stop:0 {AppStyles.PRIMARY_COLOR},
                    stop:1 {AppStyles.PRIMARY_LIGHT});
                border-radius: 12px;
            }}
            QLabel {{
                color: white;
                background: transparent;
            }}
        """)
        
        # è®¾ç½®æŒ‰é’®é¼ æ ‡æ ·å¼
        self.ui.refreshUsbBtn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.ui.refreshDriveBtn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.ui.writeTextBtn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.ui.uploadFileBtn.setCursor(Qt.CursorShape.PointingHandCursor)
        
        # éšè—è¿›åº¦ç›¸å…³æ§ä»¶
        self.ui.progressBar.setVisible(False)
        self.ui.speedLabel.setVisible(False)

        # è®¾ç½®è¡¨æ ¼åˆ—å®½è‡ªé€‚åº” - è®©æ‰€æœ‰è¡¨æ ¼çš„åˆ—å……æ»¡å®½åº¦
        self.ui.usbTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.ui.drivesTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.ui.filesTable.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
    
    def connect_signals(self):
        """è¿æ¥ä¿¡å·å’Œæ§½"""
        self.ui.refreshUsbBtn.clicked.connect(self.scan_usb_devices)
        self.ui.refreshDriveBtn.clicked.connect(self.scan_mounted_drives)
        self.ui.writeTextBtn.clicked.connect(self.write_text_file)
        self.ui.uploadFileBtn.clicked.connect(self.upload_file)
        self.ui.showHiddenCheck.stateChanged.connect(self.refresh_file_list)
        self.ui.drivesTable.itemSelectionChanged.connect(self.on_drive_selected)

    def create_table_item(self, text):
        """åˆ›å»ºä¸€ä¸ªå¸¦æœ‰å·¥å…·æç¤ºçš„è¡¨æ ¼é¡¹"""
        item_text = str(text)
        item = QTableWidgetItem(item_text)
        # è®¾ç½®å·¥å…·æç¤ºï¼Œå½“é¼ æ ‡æ‚¬åœæ—¶æ˜¾ç¤ºå®Œæ•´å†…å®¹
        item.setToolTip(item_text)
        return item
    
    def scan_usb_devices(self):
        """æ‰«æ USB è®¾å¤‡"""
        self.statusBar().showMessage("ğŸ”„ æ­£åœ¨æ‰«æ USB è®¾å¤‡...")
        devices = USBScanner.scan_devices()
        
        self.ui.usbTable.setRowCount(len(devices))
        
        for row, device in enumerate(devices):
            self.ui.usbTable.setItem(row, 0, self.create_table_item(device['name']))
            self.ui.usbTable.setItem(row, 1, self.create_table_item(device['manufacturer']))
            self.ui.usbTable.setItem(row, 2, self.create_table_item(device['serial']))
            self.ui.usbTable.setItem(row, 3, self.create_table_item(device['bus']))
            self.ui.usbTable.setItem(row, 4, self.create_table_item(device['speed']))
            self.ui.usbTable.setItem(row, 5, self.create_table_item(device['vid_pid']))
        
        self.statusBar().showMessage(f"âœ… æ‰¾åˆ° {len(devices)} ä¸ª USB è®¾å¤‡")
    
    def scan_mounted_drives(self):
        """æ‰«æå·²æŒ‚è½½çš„é©±åŠ¨å™¨"""
        self.statusBar().showMessage("ğŸ”„ æ­£åœ¨æ‰«æ U ç›˜...")
        drives = DriveManager.scan_mounted_drives()
        
        self.ui.drivesTable.setRowCount(len(drives))
        
        for row, drive in enumerate(drives):
            self.ui.drivesTable.setItem(row, 0, self.create_table_item(drive['name']))
            self.ui.drivesTable.setItem(row, 1, self.create_table_item(drive['path']))
            self.ui.drivesTable.setItem(row, 2, self.create_table_item(drive['filesystem']))
            self.ui.drivesTable.setItem(row, 3, self.create_table_item(drive['total']))
            self.ui.drivesTable.setItem(row, 4, self.create_table_item(drive['used']))
            self.ui.drivesTable.setItem(row, 5, self.create_table_item(drive['free']))
        
        self.statusBar().showMessage(f"âœ… æ‰¾åˆ° {len(drives)} ä¸ªå­˜å‚¨è®¾å¤‡")
    
    def on_drive_selected(self):
        """é©±åŠ¨å™¨é€‰ä¸­äº‹ä»¶"""
        selected_items = self.ui.drivesTable.selectedItems()
        if selected_items:
            # è·å–é€‰ä¸­è¡Œçš„ä¿¡æ¯
            row = selected_items[0].row()
            name = self.ui.drivesTable.item(row, 0).text()
            drive_path = self.ui.drivesTable.item(row, 1).text()
            
            self.selected_drive = drive_path
            self.refresh_file_list()
            
            # æ›´æ–°ç•Œé¢ä¸Šçš„é€‰ä¸­çŠ¶æ€æ ‡ç­¾
            status_text = f"ğŸ“‚ å½“å‰è®¾å¤‡: {name} ({drive_path})"
            self.ui.selectedDriveLabel1.setText(status_text)
            self.ui.selectedDriveLabel2.setText(status_text)
            self.ui.selectedDriveLabel1.setStyleSheet("color: #00695C; font-weight: bold; padding-left: 5px;")
            self.ui.selectedDriveLabel2.setStyleSheet("color: #00695C; font-weight: bold; padding-left: 5px;")
            
            self.statusBar().showMessage(f"ğŸ“ å·²é€‰æ‹©: {drive_path}")
        else:
            # æœªé€‰ä¸­ä»»ä½•è®¾å¤‡æ—¶é‡ç½®
            self.selected_drive = None
            self.ui.filesTable.setRowCount(0)
            
            reset_text = "å½“å‰è®¾å¤‡: æœªé€‰æ‹©"
            self.ui.selectedDriveLabel1.setText(reset_text)
            self.ui.selectedDriveLabel2.setText(reset_text)
            self.ui.selectedDriveLabel1.setStyleSheet("color: #666; font-weight: bold; padding-left: 5px;")
            self.ui.selectedDriveLabel2.setStyleSheet("color: #666; font-weight: bold; padding-left: 5px;")
    
    def refresh_file_list(self):
        """åˆ·æ–°æ–‡ä»¶åˆ—è¡¨"""
        if not self.selected_drive:
            return
        
        show_hidden = self.ui.showHiddenCheck.isChecked()
        files = DriveManager.list_files(self.selected_drive, show_hidden)
        
        self.ui.filesTable.setRowCount(len(files))
        
        for row, file_info in enumerate(files):
            self.ui.filesTable.setItem(row, 0, self.create_table_item(file_info['name']))
            self.ui.filesTable.setItem(row, 1, self.create_table_item(file_info['type']))
            self.ui.filesTable.setItem(row, 2, self.create_table_item(file_info['size']))
            
            # æ·»åŠ åˆ é™¤æŒ‰é’®
            if not file_info['is_dir']:
                delete_btn = QPushButton("ğŸ—‘ï¸ åˆ é™¤")
                delete_btn.setCursor(Qt.CursorShape.PointingHandCursor)
                delete_btn.clicked.connect(lambda checked, path=file_info['path']: self.delete_file(path))
                self.ui.filesTable.setCellWidget(row, 3, delete_btn)
    
    def write_text_file(self):
        """å†™å…¥æ–‡æœ¬æ–‡ä»¶"""
        if not self.selected_drive:
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·å…ˆé€‰æ‹©ä¸€ä¸ª U ç›˜")
            return
        
        filename = self.ui.filenameInput.text()
        content = self.ui.textContent.toPlainText()
        
        if not filename:
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·è¾“å…¥æ–‡ä»¶å")
            return
        
        if DriveManager.write_text_file(self.selected_drive, filename, content):
            QMessageBox.information(self, "æˆåŠŸ", f"æ–‡ä»¶ '{filename}' å†™å…¥æˆåŠŸï¼")
            self.refresh_file_list()
            self.statusBar().showMessage(f"âœ… æ–‡ä»¶ '{filename}' å†™å…¥æˆåŠŸ")
        else:
            QMessageBox.critical(self, "é”™è¯¯", "æ–‡ä»¶å†™å…¥å¤±è´¥")
            self.statusBar().showMessage("âŒ æ–‡ä»¶å†™å…¥å¤±è´¥")
    
    def upload_file(self):
        """ä¸Šä¼ æ–‡ä»¶åˆ° U ç›˜"""
        if not self.selected_drive:
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·å…ˆé€‰æ‹©ä¸€ä¸ª U ç›˜")
            return
        
        file_path, _ = QFileDialog.getOpenFileName(
            self, "é€‰æ‹©è¦ä¸Šä¼ çš„æ–‡ä»¶", "", "æ‰€æœ‰æ–‡ä»¶ (*.*)"
        )
        
        if not file_path:
            return
        
        source_path = Path(file_path)
        dest_path = Path(self.selected_drive) / source_path.name
        
        # æ˜¾ç¤ºè¿›åº¦æ¡
        self.ui.progressBar.setVisible(True)
        self.ui.speedLabel.setVisible(True)
        self.ui.progressBar.setValue(0)
        
        # åˆ›å»ºä¼ è¾“çº¿ç¨‹
        self.transfer_thread = FileTransferThread(str(source_path), str(dest_path))
        self.transfer_thread.progress.connect(self.update_progress)
        self.transfer_thread.finished.connect(self.transfer_finished)
        self.transfer_thread.start()
        
        self.statusBar().showMessage(f"ğŸ“¤ æ­£åœ¨ä¸Šä¼ : {source_path.name}")
    
    def update_progress(self, progress_percent, speed_text):
        """æ›´æ–°è¿›åº¦å’Œé€Ÿåº¦"""
        self.ui.progressBar.setValue(progress_percent)
        self.ui.speedLabel.setText(f"ä¼ è¾“é€Ÿåº¦: {speed_text}")
    
    def transfer_finished(self, success, message):
        """ä¼ è¾“å®Œæˆ"""
        self.ui.progressBar.setVisible(False)
        self.ui.speedLabel.setVisible(False)
        self.refresh_file_list()
        
        if success:
            QMessageBox.information(self, "æˆåŠŸ", message)
            self.statusBar().showMessage(f"âœ… {message}")
        else:
            QMessageBox.warning(self, "å¤±è´¥", message)
            self.statusBar().showMessage(f"âŒ {message}")
    
    def delete_file(self, file_path):
        """åˆ é™¤æ–‡ä»¶"""
        reply = QMessageBox.question(
            self, "ç¡®è®¤åˆ é™¤",
            f"ç¡®å®šè¦åˆ é™¤æ–‡ä»¶å—ï¼Ÿ\n{file_path}",
            QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No
        )
        
        if reply == QMessageBox.StandardButton.Yes:
            if DriveManager.delete_file(file_path):
                self.refresh_file_list()
                self.statusBar().showMessage(f"âœ… æ–‡ä»¶å·²åˆ é™¤: {Path(file_path).name}")
            else:
                QMessageBox.critical(self, "é”™è¯¯", "æ–‡ä»¶åˆ é™¤å¤±è´¥")
                self.statusBar().showMessage("âŒ æ–‡ä»¶åˆ é™¤å¤±è´¥")
    
    def auto_refresh(self):
        """è‡ªåŠ¨åˆ·æ–°"""
        # å¦‚æœå½“å‰åœ¨ USB è®¾å¤‡æ ‡ç­¾é¡µï¼Œåˆ·æ–° USB è®¾å¤‡
        if self.ui.tabWidget.currentIndex() == 0:
            self.scan_usb_devices()
    
    def refresh_all(self):
        """åˆ·æ–°æ‰€æœ‰æ•°æ®"""
        self.scan_usb_devices()
        self.scan_mounted_drives()